### PR TITLE
core/server: Add environment variable to write stacktrace to file

### PR DESCRIPTION
**Pull request by jasonodonnell**
When Vault receives a SIGUSR2 signal it will write a stacktrace of the running goroutines to Vault's log file. These can be hard to parse due to log prefixes and in some cases, the logger is unable to log the entire stacktrace. This adds an environment variable will force Vault to also write the stacktrace to a temporary file on the server.

### CODE REVIEW

File: command/server.go [hunk 1]
```diff
@@ -1760,6 +1760,24 @@ func (c *ServerCommand) Run(args []string) int {
 		case <-c.SigUSR2Ch:
 			logWriter := c.logger.StandardWriter(&hclog.StandardLoggerOptions{})
 			pprof.Lookup("goroutine").WriteTo(logWriter, 2)
+
+			if os.Getenv("VAULT_WRITE_STACKTRACE_TO_FILE") != "" {
+				c.logger.Info("Writing stacktrace to file..")
+
+				tmpFile, err := ioutil.TempFile("", fmt.Sprintf("%s-", filepath.Base(os.Args[0])))
```
Comment (hghaf099) [inline command/server.go:1767]: Seems like ioutil.TemFile has been deprecated. Shall we use `os.CreateTemp` instead?

**Review by hghaf099 — APPROVED** [review]
Just a nit, otherwise looks good!

File: command/server.go (hunk 1 above)
Comment (jasonodonnell) [reply command/server.go]: Good catch, fixed!

File: command/server.go [hunk 2]
```diff
@@ -1760,6 +1760,24 @@ func (c *ServerCommand) Run(args []string) int {
...
+				if err != nil {
+					c.logger.Error("Could not create temporary file", err)
+					continue
+				}
+
+				if err := pprof.Lookup("goroutine").WriteTo(tmpFile, 2); err != nil {
+					tmpFile.Close()
```
[the remark this answers is not in the document]
Comment (jasonodonnell) [reply command/server.go]: See the updated rework from earlier review 😄

File: command/server.go [hunk 3]
```diff
@@ -1760,6 +1760,24 @@ func (c *ServerCommand) Run(args []string) int {
 		case <-c.SigUSR2Ch:
 			logWriter := c.logger.StandardWriter(&hclog.StandardLoggerOptions{})
 			pprof.Lookup("goroutine").WriteTo(logWriter, 2)
+
+			if os.Getenv("VAULT_WRITE_STACKTRACE_TO_FILE") != "" {
```
[the remark this answers is not in the document]
Comment (jasonodonnell) [reply command/server.go:1764]: I can make this configurable, but there's definitely some value in also just using the OS temp directory. Some systems, like containers, have immutable filesystems unless volumes are mounted, so making an override is helpful.

File: command/server.go [hunk 4]
```diff
@@ -1760,6 +1760,32 @@ func (c *ServerCommand) Run(args []string) int {
...
+
+				f, err := os.Create(fmt.Sprintf("%s/%s", dir, "stacktrace.log"))
+				if err != nil {
+					c.logger.Error("Could not create temporary file for stacktrace", err)
+					continue
+				}
+				defer f.Close()
+
+				if err := pprof.Lookup("goroutine").WriteTo(f, 2); err != nil {
+					f.Close()
```
Comment (raskchanky) [inline command/server.go:1781]: Is this necessary given the `defer f.Close()` on line 1778?

File: command/server.go [hunk 5]
```diff
@@ -1760,6 +1760,32 @@ func (c *ServerCommand) Run(args []string) int {
...
+					c.logger.Error("Could not write stacktrace to file", err)
+					continue
+				}
+
+				c.logger.Info(fmt.Sprintf("Wrote stacktrace to: %s", f.Name()))
+				f.Close()
```
Comment (raskchanky) [inline command/server.go:1787]: Is this necessary given the `defer f.Close()` on line 1778?

File: command/server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (raskchanky) [reply command/server.go:1764]: Agreed

File: command/server.go (hunk 4 above)
Comment (raskchanky) [reply command/server.go:1790]: Ah, I see. So really, it was the defer that was out of place. Thanks for clarifying!

**Review by raskchanky — APPROVED** [review]
